VoodooPS2Trackpad/VoodooPS2ALPSGlidePoint.cpp

```diff
@@ -272,549 +415,3921 @@ bool ApplePS2ALPSGlidePoint::start( IOService * provider )
     
-    int x = (packet[1] & 0x7f) | ((packet[2] & 0x78) << (7-3));
-    int y = (packet[4] & 0x7f) | ((packet[3] & 0x70) << (7-4));
-    int z = packet[5]; // touch pression
+    // Setup expected packet size
+    priv.pktsize = priv.proto_version == ALPS_PROTO_V4 ? 8 : 6;
     
-    clock_get_uptime(&now_abs);
+    if (!(this->*hw_init)()) {
+        goto init_fail;
+    }
+    
+    return true;
     
-    left  |= (packet[2]) & 1;
-    left  |= (packet[3]) & 1;
-    right |= (packet[3] >> 1) & 1;
+init_fail:
+    IOLog("ALPS: Hardware initialization failed. TouchPad probably won't work\n");
+    resetMouse();
+    return false;
+}
 
-    if (packet[0] != 0xff)
-    {
-        left   |= (packet[0]) & 1;
-        right  |= (packet[0] >> 1) & 1;
-        middle |= (packet[0] >> 2) & 1;
-        middle |= (packet[3] >> 2) & 1;
-    }
+/* ============================================================================================== */
+/* ==============================||\\ alps.c Implementation //||================================= */
+/* ============================================================================================== */
 
-    buttons |= left ? 0x01 : 0;
-    buttons |= right ? 0x02 : 0;
-    buttons |= middle ? 0x04 : 0;
 
-    /*DEBUG_LOG("Absolute packet: x: %d, y: %d, xpos: %d, ypos: %d, buttons: %x, "
-              "z: %d, zpos: %d\n", x, y, (int)_xpos, (int)_ypos, (int)buttons, 
-              (int)z, (int)_zpos);*/
+void ApplePS2ALPSGlidePoint::alps_process_packet_v1_v2(UInt8 *packet) {
+    
+    // Check if input is disabled via ApplePS2Keyboard request
+    if (ignoreall)
+        return;
     
-    wasNotScrolling = _scrolling == SCROLL_NONE;
-    scroll = insideScrollArea(x, y);
+    int x, y, z, ges, fin, left, right, middle, buttons = 0, fingers = 0;
+    //int back = 0, forward = 0;
+    uint64_t now_abs;
     
-    willScroll = ((scroll & SCROLL_VERT) && _edgevscroll) || 
-                    ((scroll & SCROLL_HORIZ) && _edgehscroll);
-
-    // Make sure we are still relative
-    if (z == 0 || (_zpos >= 1 && z != 0 && !willScroll))
-    {
-        _xpos = x;
-        _ypos = y;
+    clock_get_uptime(&now_abs);
+    
+    if (priv.proto_version == ALPS_PROTO_V1) {
+        left = packet[2] & 0x10;
+        right = packet[2] & 0x08;
+        middle = 0;
+        x = packet[1] | ((packet[0] & 0x07) << 7);
+        y = packet[4] | ((packet[3] & 0x07) << 7);
+        z = packet[5];
+    } else {
+        left = packet[3] & 1;
+        right = packet[3] & 2;
+        middle = packet[3] & 4;
+        x = packet[1] | ((packet[2] & 0x78) << (7 - 3));
+        y = packet[4] | ((packet[3] & 0x70) << (7 - 4));
+        z = packet[5];
     }
     
-    // Are we scrolling?
-    if (willScroll)
-    {
-        if (_zscrollpos <= 0 || wasNotScrolling)
-        {
-            _xscrollpos = x;
-            _yscrollpos = y;
+    // macOS does not support forward and back buttons
```

Comment:
Hrm, are you sure? I have mouse back and forward buttons working in Safari. Need to send them through the keyboard.

Reply:
I thought of ideas for implementation:

1. Use dispatchKeyboardEventX to pass the adb codes for back and forward buttons, but I did not see any adb codes for it yet.
2. Use VoodooInput to use the 'Swipe between pages' gesture with virtual fingers:
![image](https://user-images.githubusercontent.com/28839925/154510583-ced483e0-9b2c-4339-a8e5-d9e9567d9671.png)

Nonetheless, I tend not to implement it at all since it only applies to old touchpads (pre-2010 maybe?).